### ISSUE TITLE
Avoid react exhaustive deps error

### ISSUE THREAD

**Issue by BreadAndRoses95** [nathanbuchar/react-hook-thunk-reducer#11]
Thanks for the small tool! To avoid infinite loop while using `// eslint-disable-next-line react-hooks/exhaustive-deps`, you should use useCallback, this will not recompute dispatch everytime and thus not trigger the effects that dispatch actions.

I changed a bit your code to do something like that :

```
export function useThunkReducer<S, A>(reducer: Reducer<S, A>, initialArg: S, init: (s: S) => S = s => s): [S, Dispatch<A | Thunk<S, A>>] {
    const [hookState, setHookState] = useState(init(initialArg));

    // State management.
    const state = useRef(hookState);
    const getState = useCallback(() => state.current, [])
    const setState = useCallback((newState) => {
        state.current = newState;
        setHookState(newState);
    },[]);

    // Reducer and augmented dispatcher.
    const reduce = useCallback((action) => reducer(getState(), action), [getState, reducer])
    const dispatch:Dispatch<A | Thunk<S, A>> = useCallback((action) => (
        typeof action === 'function'
            // @ts-ignore
            ? action(dispatch, getState)
            : setState(reduce(action))
    ), [getState, reduce, setState]);

    return [hookState, dispatch];
}
```

**Comment by nathanbuchar** [comment]
Haha, I think that's what I had originally but someone opened an [issue](https://github.com/nathanbuchar/react-hook-thunk-reducer/pull/1) stating the `useCallback()`s were redundant when using an empty dependency list. I'm still not totally clear on the nuances of using `useCallback()` without dependencies; know of any good resources which clears it up? I'm happy to add it back in as long as I know it's the right move

**Comment by BreadAndRoses95** [comment]
I read this article https://overreacted.io/a-complete-guide-to-useeffect when I was was first introduced to hooks but the official documentation is also clear.
I'll try to do my best to explain what I understand of the problem. Ref and the dispatcher of the useState hook are the same at every re render. (setHookState and state). As all the callbacks created after are using those "constants across re render objects", I could have used an empty dependency array, so they will only be computed at first render. Using the linter rule `react-hooks/exhaustive-deps add` add them anyway. But, not using useCallback, recretae them at every re render, which means they will change everytime state changes.

To sum it up, as you said, all of them could have been created with an empty dependency array, meaning they will not have to be updated between renders, and will keep their initial values all the time. Putting constant in dependency array just avoid linter errors. But doing this this way means that you can use the returned dispatch function in any hook dependency and be sure this will not trigger the corresponding hook (like triggering an effect every time the state is changed).

**Comment by nathanbuchar** [comment]
>  I could have used an empty dependency array, so they will only be computed at first render.

You know, that's what I originally thought! But I began second-guessing myself after both my linter yelled at me for not bothering to define dependencies, and I got the PR suggestion to remove them as redundant.

I'll look back at this in a few days and add them back in after a bit more digging for my own sanity

**Comment by verner-lall** [comment]
Any update? I think it is currently causing unnecessary rerenders on state changes in everything that uses dispatch.

**Comment by MrFatt** [comment]
> Don’t Lie to React About Dependencies

And React guarantees that both `setState` from `useState` and `dispatch` from `useReducer` are identical in each render, you can just put them into the deps list without worrying the re-render issue. 

So personally I think it would be better to wrap the dispatch with `useCallback()`.  To make sure the decorated `dispatch` behaves exactly same with the official `dispatch`

**Comment by nikparo** [comment]
> Haha, I think that's what I had originally but someone opened an [issue](https://github.com/nathanbuchar/react-hook-thunk-reducer/pull/1) stating the `useCallback()`s were redundant when using an empty dependency list. I'm still not totally clear on the nuances of using `useCallback()` without dependencies; know of any good resources which clears it up? I'm happy to add it back in as long as I know it's the right move

There's a big difference between an **empty** dependency array and **no** dependency array.

Without an array, useCallback is indeed unnecessary, as it will just pass through the function. This is what you had:
``` js
const getState = useCallback(() => state.current);
```

With an array, useCallback will only return the new function **when an element in the array has changed**. In other words, with an empty array it never updates. E.g.:
``` js
const getState = useCallback(() => state.current, []);
```
